% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/minimap2_align.R
\name{minimap2_align}
\alias{minimap2_align}
\title{Minimap2 Align to Genome}
\usage{
minimap2_align(
  config,
  fa_file,
  fq_in,
  annot,
  outdir,
  minimap2_dir,
  samtools = NULL,
  prefix = NULL,
  threads = NULL
)
}
\arguments{
\item{config}{Parsed list of FLAMES config file}

\item{fa_file}{Path to the fasta file used as a reference database for alignment}

\item{fq_in}{File path to the fastq file used as a query sequence file}

\item{annot}{Genome annotation file used to create junction bed files}

\item{outdir}{Output folder}

\item{minimap2_dir}{Path to the directory containing minimap2}

\item{samtools}{path to the samtools binary, required for large datasets since \code{Rsamtools} does not support \code{CSI} indexing}

\item{prefix}{String, the prefix (e.g. sample name) for the outputted BAM file}

\item{threads}{Integer, threads for minimap2 to use, see minimap2 documentation for details,
FLAMES will try to detect cores if this parameter is not provided.}
}
\value{
Path to the alignment file
}
\description{
Uses minimap2 to align sequences agains a reference databse.
Uses options "-ax splice -t 12 -k14 --secondary=no \code{fa_file} \code{fq_in}"
}
\examples{
temp_path <- tempfile()
bfc <- BiocFileCache::BiocFileCache(temp_path, ask = FALSE)
file_url <- "https://raw.githubusercontent.com/OliverVoogd/FLAMESData/master/data"
fastq1 <- bfc[[names(BiocFileCache::bfcadd(bfc, "Fastq1", paste(file_url, "fastq/sample1.fastq.gz", sep = "/")))]]
genome_fa <- bfc[[names(BiocFileCache::bfcadd(bfc, "genome.fa", paste(file_url, "SIRV_isoforms_multi-fasta_170612a.fasta", sep = "/")))]]
annotation <- bfc[[names(BiocFileCache::bfcadd(bfc, "annot.gtf", paste(file_url, "SIRV_isoforms_multi-fasta-annotation_C_170612a.gtf", sep = "/")))]]
outdir <- tempfile()
dir.create(outdir)
if (is.character(locate_minimap2_dir())) {
    minimap2_align(
        config = jsonlite::fromJSON(system.file("extdata/SIRV_config_default.json", package = "FLAMES")),
        fa_file = genome_fa,
        fq_in = fastq1,
        annot = annotation,
        outdir = outdir
    )
}
}
\seealso{
[minimap2_realign()]
}
