# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

find_isoform_cpp <- function(gff3, genome_bam, isoform_gff3, tss_test_stat, genomefa, transcript_fa, downsample_ratio, config_list, raw_splice_isoform) {
    .Call(`_FLAMES_find_isoform_cpp`, gff3, genome_bam, isoform_gff3, tss_test_stat, genomefa, transcript_fa, downsample_ratio, config_list, raw_splice_isoform)
}

flames_test_func <- function() {
    .Call(`_FLAMES_flames_test_func`)
}

bam_read <- function(bam_in, chr, s, e) {
    invisible(.Call(`_FLAMES_bam_read`, bam_in, chr, s, e))
}

gtf_to_bed_cpp <- function(in_gtf, out_bed, chrom_sizes_file) {
    invisible(.Call(`_FLAMES_gtf_to_bed_cpp`, in_gtf, out_bed, chrom_sizes_file))
}

#' Match Cell Barcodes
#'
#' @description Match cell barcodes in the given fastq directory with the reference csv, \code{ref_csv}. Matches are returned
#' in the output file \code{out_fastq}
#'
#' @param fastq_dir directory containing fastq files to match
#' @param stats_file NEEDED
#' @param out_fastq output filename for matched barcodes
#' @param ref_csv NEEDED
#' @param MAX_DIST int; maximum edit distance
#' @param UMI_LEN int; length of UMI sequences
#'
#' @return returns NULL
#' @import zlibbioc
#' @useDynLib FLAMES, .registration=TRUE
#' @export
match_cell_barcode <- function(fastq_dir, stats_file, out_fastq, ref_csv, MAX_DIST, UMI_LEN = 10L) {
    invisible(.Call(`_FLAMES_match_cell_barcode`, fastq_dir, stats_file, out_fastq, ref_csv, MAX_DIST, UMI_LEN))
}

#' Merge Bulk Fastq Files
#' 
#' @description Merge all fastq files into a single fastq
#' This function also inserts the original fastq file name into the header line of each read.
#'
#' @param fastq_files a string vector of the fastq file paths to merge
#' @param out_fastq the fastq file path of the output file
#' @return returns NULL
#' @useDynLib FLAMES, .registration=TRUE
#' @import zlibbioc
merge_bulk_fastq_cpp <- function(fastq_files, out_fastq) {
    invisible(.Call(`_FLAMES_merge_bulk_fastq_cpp`, fastq_files, out_fastq))
}

minimap2_align_cpp <- function(mm2_prog_path, fa_file, fq_in, sam_out, no_flank, bed12_junc) {
    invisible(.Call(`_FLAMES_minimap2_align_cpp`, mm2_prog_path, fa_file, fq_in, sam_out, no_flank, bed12_junc))
}

minimap2_tr_align_cpp <- function(mm2_prog_path, fa_file, fq_in, sam_out) {
    invisible(.Call(`_FLAMES_minimap2_tr_align_cpp`, mm2_prog_path, fa_file, fq_in, sam_out))
}

parse_gff_or_gtf_R <- function(filename) {
    .Call(`_FLAMES_parse_gff_or_gtf_R`, filename)
}

parse_json_config_cpp <- function(json_file) {
    .Call(`_FLAMES_parse_json_config_cpp`, json_file)
}

print_config_cpp <- function(list) {
    invisible(.Call(`_FLAMES_print_config_cpp`, list))
}

read_entire_bam <- function(bam_in, log_out) {
    invisible(.Call(`_FLAMES_read_entire_bam`, bam_in, log_out))
}

quantification_cpp <- function(config_list, realign_bam, transcript_fa_idx, isoform_objects_list, tr_cnt_csv, tr_badcov_cnt_csv, isoform_gff3, annot, isoform_gff3_f, FSM_anno_out) {
    invisible(.Call(`_FLAMES_quantification_cpp`, config_list, realign_bam, transcript_fa_idx, isoform_objects_list, tr_cnt_csv, tr_badcov_cnt_csv, isoform_gff3, annot, isoform_gff3_f, FSM_anno_out))
}

